File: Data Structures & Algorithms/max-area-of-island/submission-1.cpp
class Solution {
    int dirs[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int res = 0;
        for(int r = 0; r < grid.size(); ++r) {
            for(int c = 0; c < grid[0].size(); ++c) {
                if(grid[r][c]) res = max(res, dfs(grid, r, c));
            }
        }
        return res;
    }
    int dfs(vector<vector<int>>& grid, int r, int c) {
        int m = grid.size(), n = grid[0].size();
        grid[r][c] = 0;
        int res = 0;
        for(int i = 0; i < 4; ++i) {
            int x = r + dirs[i][0], y = c + dirs[i][1];
            if(x >= 0 && x < m && y >= 0 && y < n && grid[x][y] == 1)
                res += dfs(grid, x, y);
        }
        return res + 1;
    }
};

File: Data Structures & Algorithms/combination-target-sum/submission-1.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        stack = []
        amount = target
        n = len(nums)

        def dfs(i):
            nonlocal res
            nonlocal amount
            nonlocal n
            nonlocal stack

            if i == n:

                return
            elif amount < 0:

                return

            elif amount == 0:
                # print(stack)
                res.append(stack.copy())
                return
            else:
                amount -= nums[i]
                stack.append(nums[i])
                dfs(i)
                amount += nums[i]
                stack.pop()
                dfs(i + 1)

        dfs(0)

        return res

            